gateway/handler_defaults.go

```diff
@@ -83,16 +117,31 @@ func (i *handler) serveDefaults(ctx context.Context, w http.ResponseWriter, r *h
 	resolvedPath := pathMetadata.LastSegment
 	switch mc.Code(resolvedPath.Cid().Prefix().Codec) {
 	case mc.Json, mc.DagJson, mc.Cbor, mc.DagCbor:
-		blockData, ok := data.(files.File)
-		if !ok { // This should never happen
+		if bytesResponse == nil { // This should never happen
 			webError(w, fmt.Errorf("decoding error: data not a usable as a file"), http.StatusInternalServerError)
 			return false
 		}
 		logger.Debugw("serving codec", "path", contentPath)
-		return i.renderCodec(r.Context(), w, r, resolvedPath, blockData, contentPath, begin, requestedContentType)
+		return i.renderCodec(r.Context(), w, r, resolvedPath, bytesResponse, contentPath, begin, requestedContentType)
 	default:
 		logger.Debugw("serving unixfs", "path", contentPath)
-		return i.serveUnixFS(r.Context(), w, r, resolvedPath, data, pathMetadata.ContentType, contentPath, begin, logger)
+		ctx, span := spanTrace(ctx, "ServeUnixFS", trace.WithAttributes(attribute.String("path", resolvedPath.String())))
+		defer span.End()
+
+		// Handling Unixfs file
+		if bytesResponse != nil {
+			logger.Debugw("serving unixfs file", "path", contentPath)
+			return i.serveFile(ctx, w, r, resolvedPath, contentPath, bytesResponse, pathMetadata.ContentType, begin)
+		}
+
+		// Handling Unixfs directory
+		if !isDirectoryHeadRequest && dirListing == nil {
+			webError(w, fmt.Errorf("unsupported UnixFS type"), http.StatusInternalServerError)
+			return false
+		}
+
+		logger.Debugw("serving unixfs directory", "path", contentPath)
+		return i.serveDirectory(ctx, w, r, resolvedPath, contentPath, isDirectoryHeadRequest, dirListing, begin, logger)
```

Comment:
Nit: I would just invert this. `if dirListing ! = nil`, call `i.serveDirectory`. Return error if nothing matched.

Reply:
SGTM, feel free to make the change